### PR TITLE
Add null check before use variables

### PR DESCRIPTION
**Pull request by YuanYingdong**
Check ua and this->parent before further code use

### CODE REVIEW

**Comment by scw00** [comment]
IMO, I'd prefer to crashing immediately rather than hiding problem into another problem.

File: proxy/ParentSelection.cc [hunk 1]
```diff
@@ -742,7 +742,7 @@ ParentRecord::Init(matcher_line *line_info)
     unavailable_server_retry_responses = new UnavailableServerResponseCodes(nullptr);
   }
 
-  if (this->parents == nullptr && go_direct == false) {
+  if (this->parents == nullptr || go_direct == false) {
```
Comment (maskit) [inline proxy/ParentSelection.cc]: I'm not familiar with this area at all, but this change doesn't look correct.

The doc says:
> Every line in the parent.config file must contain either a parent= or go_direct= directive.

So the condition is checking either of them is configured, I think.

Comment (zwoop) [reply proxy/ParentSelection.cc]: Agreed.

Comment (jrushford) [reply proxy/ParentSelection.cc]: @maskit and @zwoop, I think this code is correct as is.  It prevents you from making a mistake in the parent.config.  If the parent list is null and you set go_direct to false any transaction matching that config would 502 as there are no parents to forward to and go_direct false prevents you from going to the origin directly.

File: proxy/http/HttpSM.cc [hunk 2]
```diff
@@ -2856,7 +2856,7 @@ HttpSM::tunnel_handler_push(int event, void *data)
   // Check to see if the client is still around
   HttpTunnelProducer *ua = (ua_txn) ? tunnel.get_producer(ua_txn) : tunnel.get_producer(HT_HTTP_CLIENT);
 
-  if (ua && !ua->read_success) {
+  if (ua == nullptr || !ua->read_success) {
```
Comment (scw00) [inline proxy/http/HttpSM.cc]: Why `ua` can be nullptr ?

**Pull request closed without merging**